### ISSUE TITLE
Help - proxy

### ISSUE THREAD

**Issue by entr0cks** [kirm/sip.js#71]
I'm somewhat confused what's the difference between sip and sip/proxy.

Would someone help out with an example to create a proxy server. 
The idea is to connect through the server to another authenticated (probably let the client deal with that) service provider (elion.ee) who enables calls between other VOIP and traditional phones. Enabling me to pick out a "call log" from INVITEs - ringing, who's calling, call end etc. So there are connections from soft-phones to the server and the server itself is connected to another server (service provider). 
Planning on connecting a html5 client as well through a websocket connection - dealing with multiple lines (notifying, putting on hold). 
Also, I don't want to pass any media through this server.

Example of calling a mobile phone 5xxxxxxx with the x-lite soft-phone, wireshark snippet: [image link](http://www.upload.ee/image/4534060/sip_login_call_busy_exit.png).

**Comment by kirm** [comment]
Sip module is a simple sip stack implementing upd/tcp/websocket transports and transaction layer as defined in rfc3261. 
Sip/proxy module is proxy server framework built on top of sip module. The main reason for its existence is tricky CANCEL
request processing rules required by protocol specification.

There are examples of proxy servers:

basic proxy-registrar server - https://github.com/kirm/sip.js/blob/master/examples/registrar-proxy.js <https://github.com/kirm/sip.js/blob/master/examples/registrar-proxy.js>
basic websocket-plain sip proxy server - https://github.com/kirm/sip.js/blob/master/examples/websocket-proxy.js <https://github.com/kirm/sip.js/blob/master/examples/websocket-proxy.js>

Kirill Mikhailov

**Comment by entr0cks** [comment]
Well, both seem to deal with the registration and client list themselves.
My scenario doesn't require that (although creating such an scenario is easy and could be used at the same time for internal communication). 
My scenario requires two connections - to client and to service provider. Passing client requests to SP and returning responses from SP to the client. Allowing me to analyse the messages (picking out INVITE's).
I'm wondering how to do this - create 2 connections, I suppose it's a possibility to create another process (cluster) which would be in a client role to communicate with the SP and use IPC between the processes.

Figure 2 on page 3 of [THIS](http://www.cisco.com/c/en/us/td/docs/ios-xml/ios/voice/cube_sipsip/configuration/12-4t/cube-sipsip-12-4t-book/voi-sip-reg-proxy.pdf) document describes my need:
![Figure 2 on page 3](http://www.upload.ee/image/4534551/cisco_sip_proxy.PNG)

**Comment by kirm** [comment]
Both examples deal with registration AND forwarding requests. You can remove registration part and implement fixed routing rules instead of lookup in registrations map.

**Comment by entr0cks** [comment]
Something's a bit off...
Used the [https://github.com/kirm/sip.js/blob/master/examples/fix302.js](https://github.com/kirm/sip.js/blob/master/examples/fix302.js) example.

![Sending to SP](http://www.upload.ee/image/4534670/sip_rec_recv.PNG)
...via.shift(); doesn't remove the previous object and the new object isn't correct either (except port and host)...

Forwarding does work (didn't realize the logic at first - sorry about that), 
SP returns 403 Forbidden. Here the ...via.shift() works.
![Receiving from SP](http://www.upload.ee/image/4534705/sip_response_rec_recv.PNG)

**Comment by kirm** [comment]
Err, I don't understand the problem.

BTW i don't recommend using fix302.js as a basis for anything that was actually a one shot hack. It has worked,  but to do the thing it was intended to do properly you need to implement so called back-2-back user agent which is much more complex than proxy server (it needs to maintain calls state not just transactions).

**Comment by entr0cks** [comment]
I'm just wondering why SP returns 403... doesn't it allow proxy's?, in which case should I forward exactly the same content only with the server's ip and port (and bypass the via object insertion, add additional code to figure out where to forward...)?
Hoped to make do with my basic knowledge, without really diving into SIP and now I'm more and more forced to learn about the standard, oh well.

**Comment by kirm** [comment]
Well if i understand correctly what you're trying to do, you need to look at proxy server example https://github.com/kirm/sip.js/blob/master/examples/registrar-proxy.js
remove special REGISTER handling (lines 16-26) and replace logic on lines 27-35 with yours.
To get proxy to work you basically need to figure out how you want to change uri of incoming request, change it and forward request (proxy.send method).

**Comment by entr0cks** [comment]
Why the uri? The messages arrive at the correct places, but the problem is that the SP returns 403 Forbidden instead of 401 Unauthorized.

I forgot to add to the pictures that the ip starting with 108 is my server and 56 is my computer's which runs x-lite (soft-phone). And 198.168.0.2 is the local address of the client...

**Comment by kirm** [comment]
What i am trying to say you should use proxy example not fix302.js and you should not do anything with via headers.

**Comment by entr0cks** [comment]
Well then it shrinks down to 
```
proxy.start({}, function (rq) {
   proxy.send(rq);
});
```
And trying to connect, server log:
recv:{ method: 'REGISTER',
send:{ uri: 'sip:elion.ee',
recv:{ method: 'REGISTER',
recv:{ method: 'REGISTER',
... (~8 requests)
send:{ uri: 'sip:elion.ee', // second sending activates now, with different branch, don't know if that's meaningful
recv:{ version: '2.0',  status: 403,
send:{ version: '2.0',  status: 403,
// Soft-phone shows "Account failed to enable ..."

**Comment by kirm** [comment]
You can respond to first request with 100 message (proxy.send(sip.makeResponse(rq, 100, 'Trying')) to prevent client to retransmitting requests.

About 403 response. Is your client able to connect to your SP directly?

**Comment by entr0cks** [comment]
Sent 100 to client, after a while forbidden is returned by SP.
Client can connect yo SP directly (first post):
![client connection directly to SP](http://www.upload.ee/image/4534060/sip_login_call_busy_exit.png)

**Comment by kirm** [comment]
Then there are two possible reasons for forbidden access. Either your
client uses different procedures to connect to SP directly (have you
compared messages?). Or you SP somehow managed to configure its servers to
deny indirect registrations.

On 22 февр. 2015 г., at 16:59, Roland Kontson <notifications@github.com>
wrote:

Sent 100 to client, after a while forbidden is returned by SP.
Client can connect yo SP directly (first post):
[image: client connection directly to SP]
<https://camo.githubusercontent.com/39210773623125d89cd6f780b5d985086ffd4fb2/687474703a2f2f7777772e75706c6f61642e65652f696d6167652f343533343036302f7369705f6c6f67696e5f63616c6c5f627573795f657869742e706e67>

—
Reply to this email directly or view it on GitHub
<https://github.com/kirm/sip.js/issues/71#issuecomment-75436607>.

**Comment by entr0cks** [comment]
Well, forwarding udp/tcp messages where I replace the ports and IPs works.

**Comment by kirm** [comment]
And which ports and ip addresses have you replaced?

On 22 февр. 2015 г., at 21:46, Roland Kontson <notifications@github.com>
wrote:

Well, forwarding udp/tcp messages where I replace the ports and IPs works.

—
Reply to this email directly or view it on GitHub
<https://github.com/kirm/sip.js/issues/71#issuecomment-75450485>.

**Comment by entr0cks** [comment]
Wow, at first I tried to replace ports and ips, but now I found that simply redirecting works too, no changing at all and everything works. Soft-phone is with it's normal configuration + my server set as proxy server and I get access to all the data, which is also broadcasted to my websocket clients and written to a text file for testing purposes.
```
var dgram = require("dgram"); //UDP
var fs = require("fs");

var remote_ip = "217.x.x.x";
var remote_port = 5060;
var counter = 0;
var socket = dgram.createSocket("udp4");

var client_ip = "84.x.x.x";
var client_port; //changes

socket.on("message", function (msg, rinfo) {
	var str = msg.toString();
	console.log("server got: " + str.substring(0, 50) + " from " + rinfo.address + ":" + rinfo.port);
	stream.write(msg); //write to file
	broadcast(JSON.stringify({ //broadcast to websocket clients
		id: ++counter,
		msg: str,
		rinfo: rinfo
	}));
	if (rinfo.address == client_ip) {
		client_port = rinfo.port;
		socket.send(msg, 0, msg.length, remote_port, remote_ip);
	} else if(rinfo.address == remote_ip) {
		socket.send(msg, 0, msg.length, client_port, client_ip);
	}
});
```

**Comment by entr0cks** [comment]
Testing between soft-phone on my Windows machine and my mobile phone (traditional).
Transmitting audio doesn't seem to work (don't have a microphone for my soft-phone).
Another issue I found was when starting a call with one side, the other sides hang-up isn't sent or received. When the side that started the call hangs up, both sides recognize it. Ringing and answering works both ways nicely.

**Comment by entr0cks** [comment]
Hanging up works only when the call is started from my mobile phone - then the BYE request comes from SP through my server.

**Comment by kirm** [comment]
I've lost any understanding of what we're trying to do here.

**Comment by entr0cks** [comment]
Yeah, well, I suppose it could be seen as a "feature request"? To transfer "raw" messages with the ability of replacing values (e.g. certain IP's and ports) instead of adding proxy routing (via..unshift()) and specifying or overriding the physical end-point (IP and port) where the packet(s) are actually sent to...
Your module is great for replying with messages, parsing and adding other extra functionality - having the capability of running an internal/private network in parallel.
Figuring out how the media stream behaves in such an "transfer"-case, correct port forwarding to the client in a port replacement scenario, the messages (like BYE) that are directly sent between clients (or client<->SP, where SP transfers to analog lines) and overriding the end-point of a sip.send() method.

**Comment by kirm** [comment]
Having re-read your messages, i think you're reinventing NAT Application Level Gateways for SIP here. 
In my experience they tend to introduce subtle, hard to diagnose problems and as a general rule my recommendation is to disable and avoid them whenever possible.
